private/bufpkg/bufplugin/bufplugindocker/docker_test.go

```diff
@@ -119,3 +186,136 @@ func buildDockerPlugin(t testing.TB, dockerClient bufplugindocker.Client, docker
 	}
 	return response, err
 }
+
+// dockerServer allows testing some failure flows by simulating the responses to Docker CLI commands.
+type dockerServer struct {
+	httpServer *httptest.Server
+	t          testing.TB
+	buildErr   error
+	pushErr    error
+	// protects builtImages
+	mu          sync.RWMutex
+	builtImages map[string]*builtImage
+}
+
+type builtImage struct {
+	tags []string
+}
+
+func newDockerServer(t testing.TB, version string) *dockerServer {
+	dockerServer := &dockerServer{t: t, builtImages: make(map[string]*builtImage)}
+	versionPrefix := "/v" + version
+	mux := http.NewServeMux()
+	mux.HandleFunc("/session", dockerServer.sessionHandler)
+	mux.HandleFunc(versionPrefix+"/build", dockerServer.buildHandler)
+	mux.HandleFunc(versionPrefix+"/images/", dockerServer.imagesHandler)
+	dockerServer.httpServer = httptest.NewServer(h2c.NewHandler(mux, &http2.Server{}))
+	t.Cleanup(dockerServer.httpServer.Close)
+	return dockerServer
+}
+
+func (d *dockerServer) sessionHandler(w http.ResponseWriter, r *http.Request) {
+	if _, err := io.Copy(io.Discard, r.Body); err != nil {
+		d.t.Error("failed to discard body:", err)
+	}
+	if r.Method != http.MethodPost {
+		http.NotFound(w, r)
+		return
+	}
+	if strings.EqualFold(r.Header.Get("Connection"), "upgrade") && r.ProtoMajor < 2 {
+		w.WriteHeader(http.StatusSwitchingProtocols)
+		return
+	}
+	w.WriteHeader(http.StatusNotFound)
+}
+
+func (d *dockerServer) buildHandler(w http.ResponseWriter, r *http.Request) {
+	if _, err := io.Copy(io.Discard, r.Body); err != nil {
+		d.t.Error("failed to discard body:", err)
+	}
+	w.WriteHeader(http.StatusOK)
+	if d.buildErr != nil {
+		d.writeError(w, d.buildErr)
+		return
+	}
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	imageID := stringid.GenerateRandomID()
+	d.builtImages[imageID] = &builtImage{tags: r.URL.Query()["t"]}
+}
+
+func (d *dockerServer) imagesHandler(w http.ResponseWriter, r *http.Request) {
+	if _, err := io.Copy(io.Discard, r.Body); err != nil {
+		d.t.Error("failed to discard body:", err)
+	}
+	d.t.Logf("%s %s", r.Method, r.URL.Path)
+	pathSuffix := strings.TrimPrefix(r.URL.Path, "/v1.41/images/")
+	imagePattern := regexp.MustCompile("^(?P<image>[^/]+/[^/]+/[^/:]+)(?::(?P<tag>[^/]+))?(?:/(?P<op>[^/]+))?$")
+	submatches := imagePattern.FindStringSubmatch(pathSuffix)
+	if len(submatches) == 0 {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+	image, tag, operation := submatches[1], submatches[2], submatches[3]
+	// ImageInspectWithRaw
+	if r.Method == http.MethodGet && operation == "json" {
+		d.mu.RLock()
+		defer d.mu.RUnlock()
+		foundImageID := d.findImageIDFromName(image + ":" + tag)
+		if len(foundImageID) == 0 {
+			http.NotFound(w, r)
+			return
+		}
+		if err := json.NewEncoder(w).Encode(&types.ImageInspect{
+			ID:       "sha256:" + foundImageID,
+			RepoTags: d.builtImages[foundImageID].tags,
```

Comment:
If we decide to use a `map[string]struct{}` instead, we can easily transform it into a slice with our `stringutil` package.